subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java

```diff
@@ -146,7 +155,15 @@ protected void configureForSourceSet(final SourceSet sourceSet, final Pmd task)
         task.setDescription("Run PMD analysis for " + sourceSet.getName() + " classes");
         task.setSource(sourceSet.getAllJava());
         ConventionMapping taskMapping = task.getConventionMapping();
-        taskMapping.map("classpath", () ->
-            sourceSet.getOutput().plus(sourceSet.getCompileClasspath()));
```

Comment:
I'm wondering if this is what we want to do. Could we add some reproducers for the common cases and what it does now/what we want it to do?

Assume project A depends on project B and project B depends on `org:foo:1.0` in some way.

I think this is the behavior with and without this change:

| dependency bucket used in B | without change | with change |
| ------------- | ------------- | ------------- |
| api  | No error | No error |
| implementation  | Error log | No error |
| compileOnly  | Error log | Error log |
| runtimeOnly  | N/A | N/A |

I think "runtimeOnly" isn't applicable here because you can't compile B against it, so it won't show up in the byte code for PMD to complain.

I'm not sure there's a way to satisfy the compileOnly dependencies from B in A without re-exposing the old "compile" behavior.

If we were to switch from compileClasspath to runtimeClasspath only, it would also fix the problem of missing B's implementation dependencies, but it would introduce a new one--missing compileOnly dependencies from A. 

I think what bothers me with this is...
1. It doesn't completely fix the problem if PMD truly needs to have a full view of the compile classpath for everything.
2. It adds another named configuration that's kind of useless for everyone else/specific to PMD.
3. The combination of compile and runtime is artificial because it doesn't represent what was used to compile and it doesn't represent what we would run with. It's close to both, but I worry about a user needing to solve resolution conflicts for something that's not something they care about. 

Is it possible that PMD only needs the compilation classpath for A and anything outside the project shouldn't be introspected in the same way? 

Or to think about it another way, would we break PMD if we started only providing interface jars for dependencies with all implementation details stripped out?

Reply:
I agree with the points you mentioned and I raised some of these concerns when we discussed the issue. 

>1. It doesn't completely fix the problem if PMD truly needs to have a full view of the compile classpath for everything.

Adding runtime classpath to PMD's auxclasspath is by no means a complete solution but it solves the `implementation` use case. I assume that it covers the vast majority of the use cases and the fix is simple enough.

>2. It adds another named configuration that's kind of useless for everyone else/specific to PMD.
I can just `plus` both compile and runtime classpath. The only difference that comes to my mind is that there might be duplicates or multiple versions of the same library. Usual JAR hell issues may arise though if different versions of dependencies are used for runtime and compile classpath :(

>3. The combination of compile and runtime is artificial because it doesn't represent what was used to compile and it doesn't represent what we would run with. It's close to both, but I worry about a user needing to solve resolution conflicts for something that's not something they care about.

I didn't think about it before but this is a big issue to consider. The configuration is "hidden" so it won't be easy to solve conflicts. It might be rare but having no way out of error is no good.

>would we break PMD if we started only providing interface jars for dependencies with all implementation details stripped out?

I've checked and the answer seems to be no because PMD is a source-level only checker. It only needs auxclasspath to construct [type definitions](https://docs.pmd-code.org/apidocs/pmd-java/6.36.0/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinition.html) that have no bytecode data available for analysis. So the interface jars should be fine but switching to them on compile classpath would be a significant project (though it'll unblock a way to other benefits).

I suppose that point 3 and the above fact are strong enough indications that we shouldn't proceed with the current PR.

Reply:
I share Sterling's understanding of the solution.
I however do not fully agree with the issues raised:
1. I agree that the solution does not cover 100% of the cases, but it is progress anyway IMO. Given the comment from the PMD contributor on #17410, the flexibility of Gradle will cause issues like this one. One way to improve this would be to add an extra bucket configuration that allows to declare extra dependencies for PMD only. As a way to pull `compileOnly` dependencies from transitive dependencies.
2. This configuration is only added when the PMD plugin is applied. So its impact makes sense. As it is _resolvable_ and _not consumable_ it will not impact other projects.
3. It is indeed artificial and there is indeed a risk of having to handle differences at some point. However experience shows that such differences are the exception and not that frequent. Also by leveraging project local resolution consistency, some of those problems can be addressed in a common way.

Reply:
We sort of got away from artificial compile + runtime mix by switching to resolving compile as runtime.